File: src/services/activeProcess.ts
import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

/**
 * Which Process is currently "active" — mirrored into Rust
 * (`active_process::ActiveProcessState`) so windows other than the main
 * one (the floating capture widget, TASK-011) can read/react to it too,
 * since Tauri windows are separate webview runtimes with no shared JS
 * memory (see docs/architecture.md, "Cross-window active-process
 * sync"). The main window's `stores/activeProcess.tsx` is the source of
 * truth for *setting* this; this service is how it's pushed to Rust and
 * how any window (including the main one, in principle) can read it
 * back or subscribe to changes.
 */
export interface ActiveProcessInfo {
  processId: string;
  processName: string;
  projectId: string;
  projectName: string;
}

const ACTIVE_PROCESS_CHANGED_EVENT = "active-process-changed";

// The IPC wire shape: matches the Rust `ActiveProcessInfo` struct's
// field names exactly (snake_case).
interface RawActiveProcessInfo {
  process_id: string;
  process_name: string;
  project_id: string;
  project_name: string;
}

function fromRaw(raw: RawActiveProcessInfo): ActiveProcessInfo {
  return {
    processId: raw.process_id,
    processName: raw.process_name,
    projectId: raw.project_id,
    projectName: raw.project_name,
  };
}

/**
 * Pushes the active Process (or `null` to clear it) to Rust, which
 * updates the system tray label and broadcasts the change to every
 * window. Fire-and-forget from the caller's point of view — a failure
 * here is logged on the backend and never needs frontend-visible error
 * handling (see docs/architecture.md).
 */
export async function syncActiveProcess(active: ActiveProcessInfo | null): Promise<void> {
  await invoke("sync_active_process", {
    input: active
      ? {
          process_id: active.processId,
          process_name: active.processName,
          project_id: active.projectId,
          project_name: active.projectName,
        }
      : null,
  });
}

/** Reads the current value directly — for a window (e.g. the widget)
 * that just opened and needs to know the active Process without waiting
 * for the next change event. */
export async function getActiveProcess(): Promise<ActiveProcessInfo | null> {
  const raw = await invoke<RawActiveProcessInfo | null>("get_active_process");
  return raw ? fromRaw(raw) : null;
}

/** Subscribes to live updates. Returns the unlisten function — callers
 * must call it on unmount to avoid leaking the listener. */
export async function onActiveProcessChanged(
  callback: (active: ActiveProcessInfo | null) => void,
): Promise<UnlistenFn> {
  return listen<RawActiveProcessInfo | null>(ACTIVE_PROCESS_CHANGED_EVENT, (event) => {
    callback(event.payload ? fromRaw(event.payload) : null);
  });
}

File: src-tauri/src/active_process.rs
//! In-memory mirror of "the active Process" (TASK-011), shared across
//! windows. The frontend's `stores/activeProcess.tsx` (in the main
//! window) is the source of truth for *setting* it, but Tauri windows
//! are separate webview processes with no shared JS memory — the
//! floating capture widget (a second window, see `tauri.conf.json`)
//! can't read the main window's React Context directly. This tiny piece
//! of Tauri-managed state, updated via
//! `commands::active_process::sync_active_process` and broadcast via an
//! app-wide event, is how the widget (and the global-shortcut handler,
//! see `hotkey.rs`) stay in sync — see docs/architecture.md,
//! "Cross-window active-process sync".
//!
//! Never persisted — resets to `None` on every launch, same as the
//! frontend store it mirrors.

use serde::{Deserialize, Serialize};
use std::sync::Mutex;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ActiveProcessInfo {
    pub process_id: String,
    pub process_name: String,
    pub project_id: String,
    pub project_name: String,
}

/// Tauri-managed state: the last value pushed by `sync_active_process`.
#[derive(Default)]
pub struct ActiveProcessState(pub Mutex<Option<ActiveProcessInfo>>);

impl ActiveProcessState {
    /// Reads the current value. Recovers from a poisoned lock (a prior
    /// holder panicking while holding it) rather than panicking itself —
    /// there is no expected path that poisons this lock, but this is
    /// cheap insurance consistent with this project's "no unwrap on a
    /// production path" convention.
    pub fn get(&self) -> Option<ActiveProcessInfo> {
        self.0.lock().unwrap_or_else(|poisoned| poisoned.into_inner()).clone()
    }

    pub fn set(&self, value: Option<ActiveProcessInfo>) {
        *self.0.lock().unwrap_or_else(|poisoned| poisoned.into_inner()) = value;
    }
}

/// Emitted (via `AppHandle::emit`, Rust-side — not gated by the frontend
/// ACL) to every window whenever the active Process changes. Payload is
/// `Option<ActiveProcessInfo>` — `null` means cleared.
pub const ACTIVE_PROCESS_CHANGED_EVENT: &str = "active-process-changed";

#[cfg(test)]
mod tests {
    use super::*;

    fn sample() -> ActiveProcessInfo {
        ActiveProcessInfo {
            process_id: "p1".to_string(),
            process_name: "Create a sales order".to_string(),
            project_id: "proj1".to_string(),
            project_name: "ERP Migration".to_string(),
        }
    }

    #[test]
    fn starts_empty() {
        let state = ActiveProcessState::default();
        assert!(state.get().is_none());
    }

    #[test]
    fn set_then_get_round_trips() {
        let state = ActiveProcessState::default();
        state.set(Some(sample()));
        assert_eq!(state.get().unwrap().process_id, "p1");
    }

    #[test]
    fn set_none_clears_a_previously_set_value() {
        let state = ActiveProcessState::default();
        state.set(Some(sample()));
        state.set(None);
        assert!(state.get().is_none());
    }
}
